File: hena/baekjoon/Implementation/14499.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define MAX 20
using namespace std;


int Array[MAX][MAX];
vector <int> commands;

int dx[4] = {0, 0, -1, 1};
int dy[4] = {1, -1, 0, 0};


bool ispossible(int x, int y, int n, int m)
{
    if (x < 0 || x >= n || y < 0 || y >= m)
        return false;
    return true;
}

void change(int *top, int *down, int *front, int * back, int * left, int * right, int dir, int x, int y)
{
    int tmp;
    if (dir == 1){
        tmp = *top;*top = *left;*left = *down;*down = *right;*right = tmp;
    }
    else if (dir == 2){
        tmp = *top;*top = *right;*right = *down;*down = *left;*left = tmp;
    }
    else if (dir == 3){
        tmp = *top;*top = *back;*back = *down;*down = *front;*front = tmp;
    }
    else if (dir == 4){
        tmp = *top;*top = *front;*front = *down;*down = *back;*back = tmp;
    }
    if (Array[x][y])
    {
        *down = Array[x][y];
        Array[x][y] = 0;
    }
    else
        Array[x][y] = *down;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    int n, m, x, y, k;
    cin >> n >> m >> x >> y >> k;

    //입력
    for (int i = 0 ; i < n ; i++)
    {
        for (int j = 0 ; j < m ; j++)
        {   
            int num;
            cin >> num;
            Array[i][j] = num;
        }
    }

    //명령어 입력
    for (int i = 0 ; i < k ; i++){
        int num;
        cin >> num;
        commands.push_back(num);
    }
    
    int top = 0, down = 0, front = 0, back = 0, left = 0, right = 0;
    for (int i = 0 ; i < k ; i++){
        int nx = x + dx[commands[i] - 1];
        int ny = y + dy[commands[i] - 1];
        if (!ispossible(nx, ny, n, m))
            continue;
        
        change(&top, &down, &front, &back, &left, &right, commands[i], nx, ny);
        cout << top << '\n';
        x = nx;
        y = ny;
    }
}

File: hena/baekjoon/Implementation/14499.py
import sys
input = sys.stdin.readline
t, d, f, b, l, r = 0, 0, 0, 0, 0, 0
dx = [0, 0, 0, -1, 1]
dy = [0, 1, -1, 0, 0]
N, M, x, y, K = map(int , input().split())


def isPossible(x, y):
    if x < 0 or x >= N or y < 0 or y >=M:
        return False
    return True

def change(dir, value, t, d, f, b, l, r, nx, ny, Map):
    # not f b
    # 동쪽
    if dir == 1:
        # r, d, l, t = t, r, d, l
        t, r, d, l = l, t, r, d,
            
    elif dir == 2:
        # r, d, l, t = d, l, t, r
        t, r, d, l = r, d, l, t

    elif dir == 3:
        t, b, d, f = f, t, b, d
    elif dir == 4:
        t, b, d, f = b, d, f, t
    if Map[nx][ny]:
        d = value
        Map[nx][ny] = 0
    else:
        Map[nx][ny] = d
    return t, d, f, b, l, r

Map = []
for _ in range(N):
    Map.append(list(map(int, input().split())))

commands = list(map(int, input().split()))

for command in commands:
    nx = x + dx[command]
    ny = y + dy[command]
    if not isPossible(nx, ny):
        continue
    
    t, d, f, b, l, r = change(command, Map[nx][ny], t, d, f, b, l, r, nx, ny, Map)
    x = nx
    y = ny
    print(t)